mixer/pkg/protobuf/yaml/dynamic/handler.go

```diff
@@ -0,0 +1,365 @@
+// Copyright 2018 Istio Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package dynamic
+
+import (
+	"context"
+	"fmt"
+	"math/rand"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/gogo/protobuf/protoc-gen-gogo/descriptor"
+	"github.com/pkg/errors"
+	"go.uber.org/atomic"
+	"google.golang.org/grpc"
+
+	"istio.io/api/mixer/adapter/model/v1beta1"
+	policypb "istio.io/api/policy/v1beta1"
+	"istio.io/istio/mixer/pkg/adapter"
+	protoyaml "istio.io/istio/mixer/pkg/protobuf/yaml"
+	istiolog "istio.io/istio/pkg/log"
+)
+
+var (
+	log = istiolog.RegisterScope("grpcAdapter", "grpc adapter debugging", 0)
+)
+
+type (
+	// Handler is the dynamic handler implementation
+	Handler struct {
+		// Name is used for debug
+		Name string
+
+		// external grpc connection
+		conn *grpc.ClientConn
+
+		connConfig *policypb.Connection
+
+		// svcMap is instance name to Svc mapping
+		svcMap map[string]*Svc
+
+		// n generates dedupeID when not given.
+		n *atomic.Uint64
+	}
+
+	// Svc encapsulates abstract service
+	Svc struct {
+		Name       string
+		Pkg        string
+		MethodName string
+		InputType  string
+		OutputType string
+
+		encoder *messageEncoder
+	}
+
+	// Instance name
+	Instance struct {
+		Name         string
+		TemplateName string
+		FileDescSet  *descriptor.FileDescriptorSet
+		Variety      v1beta1.TemplateVariety
+	}
+
+	// Codec in no-op on the way out and unmarshals using normal means
+	// on the way in.
+	Codec struct{}
+)
+
+// BuildHandler creates a dynamic handler object exposing specific handler interfaces.
+func BuildHandler(name string, connConfig *policypb.Connection, sessionBased bool, adapterConfig proto.Marshaler,
+	instances []*Instance) (hh *Handler, err error) {
+
+	hh = &Handler{
+		Name:       name,
+		svcMap:     make(map[string]*Svc, len(instances)),
+		connConfig: connConfig,
+		n:          &atomic.Uint64{},
+	}
+
+	var svc *Svc
+	for _, inst := range instances {
+		if svc, err = RemoteAdapterSvc("Handle",
+			protoyaml.NewResolver(inst.FileDescSet), sessionBased, adapterConfig); err != nil {
+			return nil, err
+		}
+		hh.svcMap[inst.Name] = svc
```

Comment:
inst.template.name

Reply:
infact there will be multiple instances of the same template.. so we need to group and dispatch after trouping.